internal/attestation/untyped.go

```diff
@@ -0,0 +1,161 @@
+		}
+		statement[predicateKey] = fixedPredicate
+	}
+
+	return untypedProvenance{statement: statement, data: embedded, signatures: signatures}, nil
+}
+
+// Detect if the value of the predicate key is a string
+func (att UntypedAttestation) predicateIsString() bool {
+	_, err := utils.MapGetStr(att, predicateKey)
+	return err == nil
+}
+
+// Attempt to extract the predicate from an arbitrary attestation, with a
+// workaround to parse a string predicate into an object if needed
+func (att UntypedAttestation) Predicate() (map[string]any, error) {
+	var predicate map[string]any
+
+	// Workaround what is (possibly) a bug in cosign where
+	// `cosign attest --predicate sbom-file.json` produces a
+	// json string version of the predicate instead of an unmarshaled
+	// object. Try to read a string first.
+	predicateStr, err := utils.MapGetStr(att, predicateKey)
+	if err == nil {
+		log.Debug("Found a string predicate. Unmarshalling it.")
+
+		// Found a string. Assume it's json and unmarshal it.
+		if err := json.Unmarshal([]byte(predicateStr), &predicate); err != nil {
+			return predicate, fmt.Errorf("Unable to unmarshal predicate string %q due to: %w", predicateStr, err)
+		}
+		return predicate, nil
+	}
+
+	// If it's not a string then it should be a map. We're expecting the format
+	// of the map could vary wildly depending on the predicateType
+	return utils.MapGetMap(att, predicateKey)
+}
+
+func (att UntypedAttestation) PredicateType() (string, error) {
+	return utils.MapGetStr(att, predicateTypeKey)
+}
+
+type untypedProvenance struct {
+	statement  UntypedAttestation
+	data       []byte
+	signatures []signature.EntitySignature
+}
+
+func (u untypedProvenance) Type() string {
+	// Maybe we don't know the predicateType, but it's still (presumably) an
+	// in-toto statement. Question: Should we extract it instead of hard coding it?
+	return in_toto.StatementInTotoV01
+}
+
+// Extract a predicateType string from the statement
+func (u untypedProvenance) PredicateType() string {
+	if t, err := u.statement.PredicateType(); err != nil {
+		panic(fmt.Errorf("Problem extracting predicateType: %w", err))
+	} else {
+		return t
+	}
+}
+
+// Because we might have done some conversion of a string predicate to an object
+// u.data might not be the same as the marshalled version of the statement. So instead
+// of returning u.data here, we'll remarshall the statement data
+func (u untypedProvenance) Statement() []byte {
+	out, err := json.Marshal(u.statement)
+	if err != nil {
+		panic(fmt.Errorf("Problem remarshaling statement: %w", err))
+	}
+	return out
+}
+
+func (u untypedProvenance) Signatures() []signature.EntitySignature {
+	return u.signatures
+}
+
+func (u untypedProvenance) MarshalJSON() ([]byte, error) {
+	val := struct {
+		Type          string                      `json:"type"`
+		PredicateType string                      `json:"predicateType"`
+		Signatures    []signature.EntitySignature `json:"signatures"`
```

Comment:
Are we missing the statement here?

Reply:
I added a todo comment to figure this out.